File: frontend/types/fee-schedule.ts
/**
 * Fee Schedule types — maps to backend FeeSchedule DTO.
 *
 * Used in:
 *   - Global defaults: /api/v1/system/fee-schedule
 *   - Merchant overrides: /api/v1/merchant-admin/merchants/:merchantId/fee-schedule
 */

/** Fee direction — must match backend FeeDirection values. */
export type FeeDirection = "processing" | "withdrawal" | "payout" | "settlement" | "refund"

export const FEE_DIRECTIONS: { value: FeeDirection; label: string; description: string }[] = [
  { value: "processing",  label: "Приём",      description: "Комиссия с входящих платежей клиентов" },
  { value: "withdrawal",  label: "Вывод",       description: "Комиссия за вывод средств мерчантом" },
  { value: "payout",      label: "Выплата",     description: "Комиссия за выплаты клиентам мерчанта" },
  { value: "settlement",  label: "Расчёт",      description: "Комиссия за взаиморасчёт" },
  { value: "refund",      label: "Возврат",     description: "Комиссия за возврат средств" },
]

/** Fee schedule entry from the API. */
export interface FeeScheduleResponse {
  merchantId: string | null
  tokenId: string
  direction: FeeDirection
  fixedFee: string       // crypto minor units as string
  percentBp: number      // basis points [0..10000]
  minFee: string         // crypto minor units
  maxFee: string         // crypto minor units
  updatedAt: string      // ISO 8601
}

/** Request body for upserting a fee schedule entry. */
export interface FeeScheduleUpsertRequest {
  tokenId: string
  direction: FeeDirection
  fixedFee: number       // int64 minor units
  percentBp: number      // basis points [0..10000]
  minFee: number
  maxFee: number
}

/** Request body for deleting a fee schedule entry. */
export interface FeeScheduleDeleteRequest {
  tokenId: string
  direction: FeeDirection
}

/** API list response. */
export interface FeeScheduleListResponse {
  items: FeeScheduleResponse[]
  total: number
}

/**
 * Formats basis points as a human-readable percentage.
 * 100 bp → "1%", 250 bp → "2.5%", 0 bp → "0%"
 */
export function formatBasisPoints(bp: number): string {
  const pct = bp / 100
  return pct % 1 === 0 ? `${pct}%` : `${pct.toFixed(2).replace(/0+$/, "")}%`
}

File: internal/infrastructure/http/v1/dto/fee_schedule.go
package dto

import "metapus/internal/domain/crypto"

// ── Fee Schedule DTOs ────────────────────────────────────────────────────

// FeeScheduleRequest is the request body for upserting a fee schedule entry.
type FeeScheduleRequest struct {
	TokenID   string `json:"tokenId" binding:"required"`
	Direction string `json:"direction" binding:"required"`
	FixedFee  int64  `json:"fixedFee"`
	PercentBP int    `json:"percentBp"`
	MinFee    int64  `json:"minFee"`
	MaxFee    int64  `json:"maxFee"`
}

// FeeScheduleResponse is the response body for a fee schedule entry.
type FeeScheduleResponse struct {
	MerchantID *string `json:"merchantId"` // nil = global default
	TokenID    string  `json:"tokenId"`
	Direction  string  `json:"direction"`
	FixedFee   string  `json:"fixedFee"`
	PercentBP  int     `json:"percentBp"`
	MinFee     string  `json:"minFee"`
	MaxFee     string  `json:"maxFee"`
	UpdatedAt  string  `json:"updatedAt"`
}

// FeeScheduleDeleteRequest is the request body for deleting a fee schedule entry.
type FeeScheduleDeleteRequest struct {
	TokenID   string `json:"tokenId" binding:"required"`
	Direction string `json:"direction" binding:"required"`
}

// FromFeeSchedule creates response DTO from domain entity.
func FromFeeSchedule(fs *crypto.FeeSchedule) FeeScheduleResponse {
	resp := FeeScheduleResponse{
		TokenID:   fs.TokenID.String(),
		Direction: string(fs.Direction),
		FixedFee:  fs.FixedFee.String(),
		PercentBP: fs.PercentBP,
		MinFee:    fs.MinFee.String(),
		MaxFee:    fs.MaxFee.String(),
		UpdatedAt: fs.UpdatedAt.Format("2006-01-02T15:04:05Z07:00"),
	}
	if fs.MerchantID != nil {
		s := fs.MerchantID.String()
		resp.MerchantID = &s
	}
	return resp
}

// FromFeeScheduleList converts a slice of domain entities to response DTOs.
func FromFeeScheduleList(schedules []crypto.FeeSchedule) []FeeScheduleResponse {
	result := make([]FeeScheduleResponse, len(schedules))
	for i := range schedules {
		result[i] = FromFeeSchedule(&schedules[i])
	}
	return result
}
